### PR TITLE
Support for Server Timing (W3C HTTP feature)

### PR DESCRIPTION
**Pull request by hendrikebbers**
See https://www.w3.org/TR/server-timing/ and https://blog.chromium.org/2018/02/chrome-65-beta-css-paint-api-and.html

The `simple-rest `sample contains a endpoint that adds some metrics:

a

b

c

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/847)

### CODE REVIEW

**Comment by hendrikebbers** [comment]
fix for https://github.com/canoo/dolphin-platform/issues/841

File: platform-examples/process-monitor-sample/process-monitor-server/src/main/java/com/canoo/dolphin/todo/server/ProcessMonitorController.java [hunk 1]
```diff
@@ -79,26 +84,31 @@ public void onDestroy() {
...
-        for (OSProcess process : procs) {
-            ProcessBean bean = null;
-            if(processList.getItems().size() <= procs.indexOf(process)) {
-                bean = beanManager.create(ProcessBean.class);
-                processList.getItems().add(bean);
-            } else {
-                bean = processList.getItems().get(procs.indexOf(process));
+        final Metric metric = serverTiming.start("processUpdate", "Update of OS process list");
+        try {
+            List<OSProcess> procs = Arrays.asList(os.getProcesses(10, OperatingSystem.ProcessSort.CPU));
```
Comment (giftkugel) [inline platform-examples/process-monitor-sample/process-monitor-server/src/main/java/com/canoo/dolphin/todo/server/ProcessMonitorController.java]: Add `final` keyword.

File: platform-examples/process-monitor-sample/process-monitor-server/src/main/java/com/canoo/dolphin/todo/server/ProcessMonitorController.java [hunk 2]
```diff
@@ -79,26 +84,31 @@ public void onDestroy() {
...
+            for (OSProcess process : procs) {
```
Comment (giftkugel) [inline platform-examples/process-monitor-sample/process-monitor-server/src/main/java/com/canoo/dolphin/todo/server/ProcessMonitorController.java]: Add `final` keyword.

File: platform-examples/process-monitor-sample/process-monitor-server/src/main/java/com/canoo/dolphin/todo/server/ProcessMonitorController.java [hunk 3]
```diff
@@ -79,26 +84,31 @@ public void onDestroy() {
...
+                bean.setProcessID(new Integer(process.getProcessID()).toString());
+                bean.setCpuPercentage(format.format(100d * (process.getKernelTime() + process.getUserTime()) / process.getUpTime()));
+                bean.setMemoryPercentage(format.format(100d * process.getResidentSetSize() / memory.getTotal()));
+                bean.setName(process.getName());
             }
-            bean.setProcessID(new Integer(process.getProcessID()).toString());
-            bean.setCpuPercentage(format.format(100d * (process.getKernelTime() + process.getUserTime()) / process.getUpTime()));
-            bean.setMemoryPercentage(format.format(100d * process.getResidentSetSize() / memory.getTotal()));
-            bean.setName(process.getName());
+            asyncServerRunner.execute(new Runnable() {
```
Comment (giftkugel) [inline platform-examples/process-monitor-sample/process-monitor-server/src/main/java/com/canoo/dolphin/todo/server/ProcessMonitorController.java]: Use Java 8 Lambda?
`asyncServerRunner.execute(() -> sessionExecutor.runLaterInClientSession(() -> update()));`

File: platform-examples/simple-rest/simple-rest-server-spring/src/main/java/com/canoo/dolphin/samples/rest/ServerApplication.java [hunk 4]
```diff
@@ -15,10 +15,10 @@
...
 
+import com.canoo.platform.server.spring.DolphinPlatformApplication;
 import org.springframework.boot.SpringApplication;
-import org.springframework.boot.autoconfigure.SpringBootApplication;
 
-@SpringBootApplication
+@DolphinPlatformApplication
 public class ServerApplication {
 
     public static void main(String[] args) {
```
Comment (giftkugel) [inline platform-examples/simple-rest/simple-rest-server-spring/src/main/java/com/canoo/dolphin/samples/rest/ServerApplication.java]: Add `final` keyword.

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/context/DolphinContext.java [hunk 5]
```diff
@@ -264,12 +266,15 @@ private void onCallControllerAction(final String controllerId, final String acti
         if (platformBeanRepository == null) {
             throw new IllegalStateException("An action was called before the init-command was sent.");
         }
+        final Metric metric = ServerTimingFilter.getCurrentTiming().start("RemotingActionCall:"+actionName, "Remote action call");
         try {
             controllerHandler.invokeAction(controllerId, actionName, params);
         } catch (Exception e) {
```
Comment (giftkugel) [inline platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/context/DolphinContext.java]: Add `final` keyword.

File: platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/servlet/DolphinPlatformServlet.java [hunk 6]
```diff
@@ -40,6 +41,11 @@ public DolphinPlatformServlet(DolphinContextCommunicationHandler communicationHa
 
     @Override
     protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
```
Comment (giftkugel) [inline platform/dolphin-platform-remoting-server/src/main/java/com/canoo/dp/impl/server/servlet/DolphinPlatformServlet.java]: Add `final` keywords.

File: platform/dolphin-platform-server/src/main/java/com/canoo/dp/impl/server/bootstrap/modules/ServerTimingModule.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        return SERVER_TIMING_MODULE_ACTIVE_PROPERTY_NAME;
+    }
+
+    @Override
+    public String getName() {
+        return MODULE_NAME;
+    }
+
+    @Override
+    public void initialize(ServerCoreComponents coreComponents) throws ModuleInitializationException {
```
Comment (giftkugel) [inline platform/dolphin-platform-server/src/main/java/com/canoo/dp/impl/server/bootstrap/modules/ServerTimingModule.java]: Add `final` keyword.

File: platform/dolphin-platform-server/src/main/java/com/canoo/dp/impl/server/servlet/GenericResponseWrapper.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PrintWriter;
+
+public class GenericResponseWrapper extends HttpServletResponseWrapper {
+    private final ByteArrayOutputStream output;
+
+    public GenericResponseWrapper(HttpServletResponse response) {
```
Comment (giftkugel) [inline platform/dolphin-platform-server/src/main/java/com/canoo/dp/impl/server/servlet/GenericResponseWrapper.java]: Add `final` keyword.

File: platform-examples/simple-rest/simple-rest-server-spring/src/main/java/com/canoo/dolphin/samples/rest/CityEndpoint.java [hunk 9]
```diff
@@ -24,12 +27,45 @@
...
+            final CityDetails cityDetails = new CityDetails(city);
+            cityDetails.setDescription("No description");
+            cityDetails.setPopulation((long) (Math.random() * 1_000_000));
+            return cityDetails;
+        } finally {
+            metric.stop();
+        }
+    }
+
+    private void sleep(long ms) {
```
Comment (giftkugel) [inline platform-examples/simple-rest/simple-rest-server-spring/src/main/java/com/canoo/dolphin/samples/rest/CityEndpoint.java]: Add `final` keyword.

File: platform-examples/process-monitor-sample/process-monitor-server/src/main/java/com/canoo/dolphin/todo/server/ProcessMonitorController.java (hunk 3 above)
Comment (giftkugel) [reply platform-examples/process-monitor-sample/process-monitor-server/src/main/java/com/canoo/dolphin/todo/server/ProcessMonitorController.java]: Or even better with `asyncServerRunner.execute(() -> sessionExecutor.runLaterInClientSession(this::update));`

Comment (hendrikebbers) [reply platform-examples/process-monitor-sample/process-monitor-server/src/main/java/com/canoo/dolphin/todo/server/ProcessMonitorController.java]: 👍

**Pull request merged**